include/pybind11/pybind11.h

```diff
@@ -2132,13 +2121,21 @@ class gil_scoped_acquire {
                     pybind11_fail("scoped_acquire::dec_ref(): internal error!");
             #endif
             PyThreadState_Clear(tstate);
-            if (!detail::is_finalizing())
+            if (active)
                 PyThreadState_DeleteCurrent();
             PYBIND11_TLS_DELETE_VALUE(detail::get_internals().tstate);
             release = false;
         }
     }
 
+    /// This method will disable the PyThreadState_DeleteCurrent call. This
+    /// should be called if the interpreter is shutting down, as the thread
+    /// deletion is not allowed during shutdown.
+    /// (_Py_IsFinalizing() on Python 3.7+)
```

Comment:
I feel like this comment is too friendly and thus doesn't really convey the dangers of disarming this RAII object.

Reply:
```suggestion
    /// This method will disable the PyThreadState_DeleteCurrent call. This
    /// should be called if the interpreter is shutting down, as the thread
    /// deletion is not allowed during shutdown. Be sure you do not try to
    /// rely on this guard if you have disarmed it!
    /// (Use _Py_IsFinalizing() on Python 3.7+ to detect shutdown)
```

Reply:
?

Reply:
Sound good, yes.
The one thing that's missing is that it's not entirely clear why we don't do the check ourselves (cfr. this original PR). But if @bstaletic is happy, this is great to me! :-)

Reply:
How... How did I not reply to this one??

I was thinking more of an explicit warning that, if disarmed, the RAII guard won't acquire the GIL. Also the `should be called` can lead to confusion whether a user *should* actually call `disarm()` anyway. Here's my proposal that I apparently forgot to post:

```
This method will disable the PyThreadState_DeleteCurrent call and the GIL won't be acquired. This
may be called if the interpreter is shutting down, as the thread
deletion is not allowed during shutdown.
```

To summarize:

1. Mention the GIL explicitly.
2. Use `may` instead of `should`.
3. Optionally restate the warning at the end of the comment, like you already proposed.
4. Optionally mention this issue discussion, like @YannickJadoul proposed, though we do have `git blame` already.
5. Don't forget the other similar comment below.

Anyway, I'm not going to nag too much about the exact wording. I just want to drive home the point that this `disarm()` is dangerous so that any future user knows what he's getting into. All that said, I'll approve

Reply:
>     4\. Optionally mention this issue discussion, like @YannickJadoul proposed, though we do have `git blame` already.

For the record, I didn't really mean this; I'm not a huge fan of (too many) links to GitHub in the code itself. I mostly meant the current explanation potentially leaves open the question on "why did you not implement this in the `gil_scoped_acquire` destructor?", since it doesn't show doing it during shutdown is still reasonably dangerous, depending on the situation;

Reply:
I'll modify just a bit, but that's about right. I do think it is "Should" - you really should do this or you will get a segfault. The reason we don't do it by default is it is _also_ dangerous to assume this is always armed; if you have a `if ... .disarm()`, then hopefully a user will also check the same condition instead assuming the guard is always armed.

In short: not calling disarm if you know the interpreter may be shutting down when this is called is also dangerous.

How about this:

> This method will disable the PyThreadState_DeleteCurrent call and the GIL won't be acquired. This
  method should be used if the interpreter could be shutting down when this is called, as the thread
  deletion is not allowed during shutdown. Check _Py_IsFinalizing() on Python 3.7+, and protect subsequent code.

Reply:
That sounds good to me.

Reply:
Great! :-)